Enforce using the compiler daemon for toolchains

bmuskalla opened the pull request:
This adds general support for using the compile daemon
over dropping down to the cli compiler if possible.
Most of the changes are about removing the assumption
that the current JVM is the worker/compiler default.

Fixes #12513

ljacomet commented on subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java:
```diff
@@ -319,7 +320,11 @@ protected ProjectLayout getProjectLayout() {
     @Deprecated
     protected JavaPlatform getPlatform() {
-        return new DefaultJavaPlatform(JavaVersion.toVersion(getTargetCompatibility()));
+        Object targetCompatibility = getTargetCompatibility();
```
I do not get this change. Tbh I do not get the code that was there before either 😉 
Can you explain what this fixes / improves?

bmuskalla replied:
As discussed, this was only needed under the previous assumption of supporting non-toolchain compiler daemons. Reverted.

ljacomet reviewed (commented):
Looks good overall and nice to see the code getting back to what it was before toolchains. Gives more confidence that we will not break some use cases out there1

One question though, see below:

ljacomet reviewed (approved):
Let's go and merge. We can resolve the question in a follow up (if needed)

ljacomet commented:
@bmuskalla sorry for the conflicts 😞

bmuskalla commented:
@ljacomet Removed some (now) unnecessary changes and reverted a refactoring that was not required anymore. Given this was now rebased onto `release` and had a number of merge conflicts, it would be great to have another set of eyes on this.

melix commented on subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DaemonJavaCompiler.java:
```diff
@@ -69,6 +71,16 @@ protected DaemonForkOptions toDaemonForkOptions(JavaCompileSpec spec) {
+        if (forkOptions.getExecutable() != null) {
+            return new File(forkOptions.getExecutable());
+        } else if (forkOptions.getJavaHome() != null) {
+            return new File(forkOptions.getJavaHome(), "bin/java");
```
:o: You probably need to use `org.gradle.internal.os.OperatingSystem#getExecutableName` here.

ljacomet commented on subprojects/core/src/main/java/org/gradle/process/internal/DefaultExecActionFactory.java:
```diff
@@ -123,9 +123,7 @@ public JavaForkOptionsInternal newDecoratedJavaForkOptions() {
-        final DefaultJavaForkOptions forkOptions = new DefaultJavaForkOptions(fileResolver, fileCollectionFactory, new DefaultJavaDebugOptions());
-        forkOptions.setExecutable(Jvm.current().getJavaExecutable());
-        return forkOptions;
+        return new DefaultJavaForkOptions(fileResolver, fileCollectionFactory, new DefaultJavaDebugOptions());
```
Reverting this change but not the one to `newDecoratedJavaForkOptions`. Is there a specific reason?

ljacomet commented on subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcessBuilder.java:
```diff
@@ -70,6 +71,7 @@
 
     DefaultWorkerProcessBuilder(JavaExecHandleFactory execHandleFactory, MessagingServer server, IdGenerator<Long> idGenerator, ApplicationClassesInSystemClassLoaderWorkerImplementationFactory workerImplementationFactory, OutputEventListener outputEventListener, MemoryManager memoryManager) {
         this.javaCommand = execHandleFactory.newJavaExec();
+        this.javaCommand.setExecutable(Jvm.current().getJavaExecutable());
```
If this was not needed before, does that mean we are shipping a change in how `JavaExecHandleFactory` behaves?
I know this is `internal`, so trying to assess if the change there is on purpose.

ljacomet reviewed (commented):
A couple questions, see below

ljacomet commented on subprojects/plugins/src/test/groovy/org/gradle/api/plugins/JavaBasePluginTest.groovy:
```diff
@@ -221,10 +221,23 @@ class JavaBasePluginTest extends AbstractProjectBuilderSpec {
+        setupProjectWithToolchain(Jvm.current().javaVersion)
+
+        when:
+        project.sourceSets.create('custom')
```
This looks strange in the test. What's the purpose?

ljacomet reviewed (approved):
Great!

One comment on a test, see below

bmuskalla commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":688757790,"teamCityBuildId":"37911564","teamCityBuildHeadRef":"fe8e3a4d9c7361bfa2e98c368c38be013e996c94","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
            
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=37911564) for you.
